src/marqo/core/typed_vespa_index.py

```diff
@@ -0,0 +1,271 @@
+            schema.append(f'expression: {bm25_expression}')
+            schema.append('}}')
+
+        if tensor_fields:
+            schema.append(f'rank-profile {cls._EMBEDDING_SIMILARITY_RANK_PROFILE} inherits default {{')
+            schema.append('inputs {')
+            schema.append(f'query({cls._RANK_INPUT_QUERY_EMBEDDING}) tensor<float>(x[{model_dim}])')
+            for field in tensor_fields:
+                schema.append(f'query({field}): 1')
+
+            schema.append('}')
+            schema.append('first-phase {')
+            schema.append(f'expression: {embedding_similarity_expression}')
+            schema.append('}}')
+
+        if score_modifier_fields:
+            expression = f'if (count(query({cls._RANK_INPUT_MULT_WEIGHTS})) == 0, 1, ' \
+                         f'reduce(query({cls._RANK_INPUT_MULT_WEIGHTS}) ' \
+                         f'* attribute({cls._SCORE_MODIFIERS_FIELD}), prod)) * score ' \
+                         f'+ reduce(query({cls._RANK_INPUT_ADD_WEIGHTS}) ' \
+                         f'* attribute({cls._SCORE_MODIFIERS_FIELD}), sum)'
+            schema.append(f'rank-profile {cls._MODIFIERS_RANK_PROFILE} inherits default {{')
+            schema.append('inputs {')
+            schema.append(f'query({cls._RANK_INPUT_MULT_WEIGHTS})  tensor<float>(p{{}})')
+            schema.append(f'query({cls._RANK_INPUT_ADD_WEIGHTS})  tensor<float>(p{{}})')
+            schema.append('}')
+            schema.append('function modify(score) {')
+            schema.append(f'expression: {expression}')
+            schema.append('}}')
+
+            if lexical_fields:
+                schema.append(f'rank-profile {cls._BM25_RANK_PROFILE}_{cls._MODIFIERS_RANK_PROFILE} '
+                              f'inherits {cls._MODIFIERS_RANK_PROFILE} {{ first-phase {{')
+                schema.append(f'expression: modify({bm25_expression})')
+                schema.append('}}')
+
+            if tensor_fields:
+                schema.append(f'rank-profile {cls._EMBEDDING_SIMILARITY_RANK_PROFILE}_{cls._MODIFIERS_RANK_PROFILE} '
+                              f'inherits {cls._MODIFIERS_RANK_PROFILE} {{ first-phase {{')
+                schema.append(f'expression: modify({embedding_similarity_expression})')
+                schema.append('}}')
+
+    @classmethod
+    def _get_vespa_type(cls, marqo_type: FieldType) -> str:
+        try:
+            return cls._TYPE_MAP[marqo_type]
+        except KeyError:
+            raise ValueError(f'Unknown Marqo type: {marqo_type}')
+
+    @classmethod
+    def _get_distance_metric(cls, marqo_distance_metric: DistanceMetric) -> str:
+        try:
+            return cls._DISTANCE_METRIC_MAP[marqo_distance_metric]
+        except KeyError:
+            raise ValueError(f'Unknown Marqo distance metric: {marqo_distance_metric}')
+
+    @classmethod
+    def _get_model_dimension(cls, marqo_index: MarqoIndex) -> int:
```

Comment:
nit: I don't feel that this logic belongs to this class. I feel that it makes more sense being part of the MarqoIndex class

Reply:
You are right. Though I don't think it belongs in `MarqoIndex` either. I don't like to have heavy logic in models. For one thing, it increases the chance of hitting circular dependencies (e.g. if model registry needs `MarqoIndex` at some point). 

This operation should be in a separate module that takes care of all model logic. This will be a refactored version of `model_registry` and pull all related functions from `s2_search` and other places into once place. 

I've added a TODO for this

Reply:
makes sense